### PR TITLE
Programs

### PR DESCRIPTION
**Pull request by tjukes**
Draft for review & discussion. A few points in particular for attention: 
- Sensible defaults for gift card programs and promotions: how much configurability should we expose initially? 
- What restrictions should we place on updating Programs? Which fields and under what circumstances? 
- What attributes should we allow Programs to be filtered on? Dates have been removed, should others be removed as well? 
- Need consistency: do Accounts use a default Program whose ID comes back in the response, or do they come back with `program: null`?

The question of issuing Values and retrieving them by `issuance` or `tags` still needs discussion before that section can be written up, so it's not included here. 

---------
Updated with changes discussed.

### CODE REVIEW

File: apiary-sources/endpoints/programs-promotions-delete.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    
+    + Body
+
+            {
+                "success": true
+            }
+
++ Response 409 (application/json)
+
+    Attempting to delete a Promotion Program that is in use. If the Program is referenced by one or more Values, those objects would need to be deleted first. If a Value created through this Program is referenced by a Transaction, it cannot be deleted because Transactions cannot be deleted. Programs in use can still be updated.
```
Comment (Tim-Jordison) [inline apiary-sources/endpoints/programs-promotions-delete.md]: Should delete a program really be a thing? To delete it means we need to query Values and Transactions. These are both ever growing lists.

File: apiary-sources/endpoints/programs-promotions-update.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
+### Update a Promotion Program [PATCH /programs/promotions/{id]
```
Comment (Tim-Jordison) [inline apiary-sources/endpoints/programs-promotions-update.md]: I don't really understand why Get Program is done via `/programs/{id}` but patching is done via `/programs/promotions/{id}`. This feels inconsistent. 

Also, there is a closing bracket '}' missing after id in the url.

File: apiary-sources/endpoints/programs-get.md [hunk 3]
```diff
@@ -1,4 +1,4 @@
-### Get Programs [GET /programs{?limit}{?offset}{?programId}{?currency}{?access}{?discount}{?preTax}{?active}{?minUses}{?maxUses}{?tags}{?minCreatedDate}{?maxCreatedDate}{?minUpdatedDate}{?maxUpdatedDate}]
+### Get Programs [GET /programs{?limit}{?offset}{?id}{?tags}{?name}{?type}{?currency}{?access}{?discount}{?preTax}{?active}{?minUses}{?maxUses}]
```
Comment (pushplay) [inline apiary-sources/endpoints/programs-get.md]: {?limit}{?offset}{?id}{?tags}{?currency}

File: apiary-sources/endpoints/programs-promotions-update.md (hunk 2 above)
Comment (pushplay) [inline apiary-sources/endpoints/programs-promotions-update.md]: `{id]` -> `{id}`

File: apiary-sources/endpoints/programs-get.md [hunk 4]
```diff
@@ -55,31 +53,33 @@
...
                     "metadata": null,
+                    "startDate": null,
+                    "endDate": null,
                     "createdDate": "2018-04-17T23:20:08.404Z",
                     "updatedDate": "2018-04-17T23:20:08.404Z"
                 }
             ]
 
-### Get Program [GET /programs/{programId}]
+### Get Program [GET /programs/{id}]
```
Comment (pushplay) [inline apiary-sources/endpoints/programs-get.md]: Delete me.

File: apiary-sources/endpoints/programs-get.md [hunk 5]
```diff
@@ -8,19 +8,17 @@
 + Parameter
     + limit (number, optional) - {{pagination.limit}}
     + offset (number, optional) - {{pagination.offset}}
-    + programId (string, optional) - Filter by programId.
+    + id (string, optional) - Filter by ID.
```
Comment (pushplay) [inline apiary-sources/endpoints/programs-get.md]: Take advantage of `{{filter.tags}}  {{filter.list}}` and `{{filter.id}}  {{filter.list}}`

File: apiary-sources/api-sections/programs.md [hunk 6]
```diff
@@ -1,18 +1,12 @@
...
+Programs are most commonly used to define and organize Promotion campaigns. For example, the `Value` issued for a new user promotion that gives them a 5% discount on their first purchase should behave differently from the `Value` for a "TEN-DOLLAR-TUESDAYS" promotion. These different promotions will also need to be tracked and analyzed separately, all of which can happen in the web app.
 
----
-{% include 'endpoints/programs-create.md' %}
-
-{% include 'endpoints/programs-update.md' %}
+Programs can also be used to control and organize gift card campaigns. Lightrail supplies you with a default gift card Program, but you can also choose to create more gift card Programs for custom use cases -- for example, so that B2B gift cards can be handled differently from gift cards that your customers purchase directly from your site.
 
-{% include 'endpoints/programs-delete.md' %}
+Behind the scenes, all Accounts are also backed by a default Accounts Program. Lightrail recommends using this default for all Accounts.
```
Comment (Tim-Jordison) [inline apiary-sources/api-sections/programs.md]: Isn't there just a null program for accounts?

File: apiary-sources/api-sections/giftcards.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+{% include 'endpoints/values-giftcards-create.md' %}
+{% include 'endpoints/values-giftcards-get.md' %}
+{% include 'endpoints/values-giftcards-update.md' %}
+{% include 'endpoints/values-giftcards-claim.md' %}
+{% include 'endpoints/values-giftcards-changeCode.md' %}
+
+### Giftcard Programs [/programs/giftcards]
+
+Programs define the behaviour of Value that you wish to issue: in this case, gift cards. Programs are typically created through the web app, but an overview of the Programs endpoints related to gift cards is included here, and a more detailed explanation of Programs can be found below to provide context and insight into how `Values` function.
```
Comment (pushplay) [inline apiary-sources/api-sections/giftcards.md]: This paragraph feels a bit lost.  Refocus on what the developer needs to know.

> an overview of the Programs endpoints related to gift cards is included here
Yeah.  That's implied by the section.

> a more detailed explanation of Programs can be found below
Where?

File: apiary-sources/api-sections/giftCards.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+{% include 'endpoints/values-giftCards-create.md' %}
+{% include 'endpoints/values-giftCards-get.md' %}
+{% include 'endpoints/values-giftCards-update.md' %}
+{% include 'endpoints/values-giftCards-claim.md' %}
+{% include 'endpoints/values-giftCards-changeCode.md' %}
+
+### Gift Card Programs [/programs/giftCards]
+
+Programs define the behaviour of Value that you wish to issue: in this case, gift cards. Programs are typically created through the web app, but an overview of the Programs endpoints related to gift cards is included here, and a more detailed explanation of Programs can be found below to provide context and insight into how `Values` function.
```
Comment (pushplay) [inline apiary-sources/api-sections/giftCards.md]: This paragraph feels a bit lost.  Refocus on what the developer needs to know.

> an overview of the Programs endpoints related to gift cards is included here

Yeah.  That's implied by the section.

> a more detailed explanation of Programs can be found below

Where?

File: apiary-sources/data-structures/Program.md [hunk 9]
```diff
@@ -14,8 +15,10 @@
 + minInitialBalance (number, optional) - {{program.minBalance}}
 + maxInitialBalance (number, optional) - {{program.maxBalance}}
 + fixedInitialValues (array[number], optional) - A list of values the Value can be created with.
```
Comment (pushplay) [inline apiary-sources/data-structures/Program.md]: Should be `fixedInitialBalances` - A list of *balances` the Value can be created with.

File: apiary-sources/api-sections/giftCards.md (hunk 8 above)
Comment (tjukes) [reply apiary-sources/api-sections/giftCards.md]: Yeah it needs a link.

Comment (tjukes) [reply apiary-sources/api-sections/giftCards.md]: Updated. Linking seems a bit weird, will likely need to fix eventually.

File: apiary-sources/api-sections/giftCards.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+{% include 'endpoints/values-giftCards-create.md' %}
+{% include 'endpoints/values-giftCards-get.md' %}
+{% include 'endpoints/values-giftCards-update.md' %}
+{% include 'endpoints/values-giftCards-claim.md' %}
+{% include 'endpoints/values-giftCards-changeCode.md' %}
+
+### Gift Card Programs [/programs/giftCards]
+
+Programs define the behaviour of Value that you wish to issue: in this case, gift cards. Programs are typically created through the web app, but an explanation of Programs can be found [below](#advanced-programs) to provide context and insight into how `Values` function. The following endpoints can be used to manage Gift Card Programs.
```
Comment (pushplay) [inline apiary-sources/api-sections/giftCards.md]: I still think this paragraph doesn't answer the fundamental question: what is a Gift Card Program?  What problem does it solve?

> The following endpoints can be used to manage Gift Card Programs. 
Yup...

**Pull request merged**